File: src/utils/unitOutputs.js
// Values are useful delivered output rates, never installed capacity or tank level.
export function unitOutputs(unit) {
  const online = ["online", "operational"].includes(unit.status);
  const definitions = {
    power: [true, unit.currentPower ?? unit.current_power, "kW"],
    heat: [unit.supports_heat, unit.usefulHeat ?? unit.useful_heat_kw, "kWth"],
    chill: [
      unit.supports_chill,
      unit.usefulChill ?? unit.useful_chill_kw,
      "kWth",
    ],
    water: [
      unit.supports_water ?? unit.watergeneration,
      unit.waterRate ?? unit.water_rate_lph,
      "L/h",
    ],
  };
  return Object.fromEntries(
    Object.entries(definitions).map(([key, [capability, reading, units]]) => {
      const source = unit.outputs?.[key];
      const value = source?.value ?? reading;
      const valid = value != null && Number.isFinite(Number(value));
      const capable = source?.capable ?? Boolean(capability);
      const stale = source?.stale ?? unit.source !== "demo";
      const quality =
        source?.quality ?? (unit.source === "demo" ? "GOOD" : "UNKNOWN");
      return [
        key,
        {
          capable,
          value: valid ? Number(value) : null,
          unit: units,
          measuredAt: source?.measuredAt ?? null,
          quality,
          stale,
          active:
            online &&
            capable &&
            valid &&
            Number(value) > 0 &&
            !stale &&
            quality === "GOOD",
        },
      ];
    }),
  );
}

File: backend/app/services/unit_outputs.py
"""Useful output contract; each channel carries its own measurement quality/age."""

import math
from datetime import datetime, timezone
from app.models import Sensor, SensorReading

CONTRACT = {
    "power": (("current_power", "power"), None, "kW", {"kw": 1, "w": 0.001}),
    "heat": (("useful_heat_kw",), "supports_heat", "kWth", {"kw": 1, "w": 0.001}),
    "chill": (("useful_chill_kw",), "supports_chill", "kWth", {"kw": 1, "w": 0.001}),
    "water": (
        ("water_flow",),
        "supports_water",
        "L/h",
        {"l/h": 1, "l/min": 60, "m3/h": 1000},
    ),
}


def output_states(unit, now=None):
    now = now or datetime.now(timezone.utc)
    online = getattr(unit.status, "value", unit.status) == "online"
    result = {}
    for key, (channels, capability, units, scales) in CONTRACT.items():
        row = (
            SensorReading.query.join(Sensor)
            .filter(
                Sensor.unit_id == unit.id,
                Sensor.is_active.is_(True),
                Sensor.sensor_type.in_(channels),
            )
            .order_by(SensorReading.timestamp.desc(), SensorReading.id.desc())
            .first()
        )
        capable = (
            capability is None
            or bool(getattr(unit, capability, False))
            or row is not None
        )
        value, measured, quality, stale = None, None, "UNKNOWN", True
        if row:
            scale = scales.get(
                (row.sensor.unit_of_measurement or "").lower().replace(" ", "")
            )
            time = (
                row.timestamp.replace(tzinfo=timezone.utc)
                if row.timestamp.tzinfo is None
                else row.timestamp
            )
            measured, quality = time.isoformat(), row.quality
            stale = not 0 <= (now - time).total_seconds() <= 900
            if scale is not None and math.isfinite(row.value):
                value = row.value * scale
            else:
                quality = "BAD"
        result[key] = {
            "capable": capable,
            "value": value,
            "unit": units,
            "measuredAt": measured,
            "quality": quality,
            "stale": stale,
            "active": online
            and capable
            and value is not None
            and value > 0
            and quality == "GOOD"
            and not stale,
        }
    return result
